opamp/observiq/agent_config_manager.go

```diff
@@ -0,0 +1,167 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package observiq
+
+import (
+	"bytes"
+	"fmt"
+	"os"
+	"path/filepath"
+
+	"github.com/observiq/observiq-otel-collector/opamp"
+	"github.com/open-telemetry/opamp-go/protobufs"
+	"go.uber.org/zap"
+)
+
+const (
+	// CollectorConfigName is the key of the collector config in OpAmp
+	CollectorConfigName = "collector.yaml"
+	// ManagerConfigName is the key of the manager config in OpAmp
+	ManagerConfigName = "manager.yaml"
+	// LoggingConfigName is the key of the logging config in OpAmp
+	LoggingConfigName = "logging.yaml"
+)
+
+// acceptableConfigs is a lookup of configs that are able to be written/updated
+var acceptableConfigs = map[string]struct{}{
+	CollectorConfigName: {},
+	ManagerConfigName:   {},
+	LoggingConfigName:   {},
+}
+
+// Enforce interface
+var _ opamp.ConfigManager = (*AgentConfigManager)(nil)
+
+// AgentConfigManager keeps track of active configs for the agent
+type AgentConfigManager struct {
+	configMap map[string]*opamp.ManagedConfig
+	logger    *zap.SugaredLogger
+}
+
+// NewAgentConfigManager creates a new AgentConfigManager
+func NewAgentConfigManager(defaultLogger *zap.SugaredLogger) *AgentConfigManager {
+	return &AgentConfigManager{
+		configMap: make(map[string]*opamp.ManagedConfig),
+		logger:    defaultLogger.Named("config manager"),
+	}
+}
+
+// AddConfig adds a config to be tracked by the config manager.
+// If the config already is tracked it'll be overwritten with the new managed config
+func (a *AgentConfigManager) AddConfig(configName string, managedConfig *opamp.ManagedConfig) {
+	a.configMap[configName] = managedConfig
+}
+
+// ComposeEffectiveConfig reads in all config files and calculates the effective config
+func (a *AgentConfigManager) ComposeEffectiveConfig() (*protobufs.EffectiveConfig, error) {
+	contentMap := make(map[string]*protobufs.AgentConfigFile, len(a.configMap))
+
+	// Used to track config names to alphabetize later in hash compution
+	configNames := make([]string, 0, len(a.configMap))
+
+	for configName, managedConfig := range a.configMap {
+		// Read in config file
+		cleanPath := filepath.Clean(managedConfig.ConfigPath)
+		configContents, err := os.ReadFile(cleanPath)
+		if err != nil {
+			return nil, fmt.Errorf("error reading config file %s: %w", configName, err)
+		}
+
+		// Add to contentMap
+		contentMap[configName] = &protobufs.AgentConfigFile{
+			Body:        configContents,
+			ContentType: opamp.DetermineContentType(managedConfig.ConfigPath),
+		}
+
+		configNames = append(configNames, configName)
+	}
+
+	return &protobufs.EffectiveConfig{
+		ConfigMap: &protobufs.AgentConfigMap{
+			ConfigMap: contentMap,
+		},
+	}, nil
+}
+
+// ApplyConfigChanges compares the remoteConfig to the existing and applies changes
+func (a *AgentConfigManager) ApplyConfigChanges(remoteConfig *protobufs.AgentRemoteConfig) (effectiveConfig *protobufs.EffectiveConfig, changed bool, err error) {
```

Comment:
minor: this method's a little large. would be nice if we could refactor to break it up a bit

Reply:
I'll factor out the new file addition. I'd like to keep the actual config change logic inline though.

Reply:
looks better. i'd still probably factor out the stuff in the for loop into a method like `compareRemoteConfig` or something as the method still seems a little large but i won't nitpick on it